### ISSUE TITLE
How do sites serve origin policies selectively?

### ISSUE THREAD

**Issue by domenic** [mikewest/origin-policy#51]
The new [update and version negotiation design](https://github.com/WICG/origin-policy/blob/master/version-negotiation.md) removed the HTTP request header, `Sec-Origin-Policy`, as I thought it was not needed with that design. However, that may have been a mistake, as it removes request-time feature detection.

In particular, what is the deployment story for sites that want to support both browsers that implement origin policy, and browsers that do not implement origin policy? With the old design (even with the variant implemented in Chrome that always sent `Sec-Origin-Policy: 0`), the flow was to indicate the origin policy with the response when the request contains `Sec-Origin-Policy`. Otherwise, you would send the current complement of headers (i.e. CSP, Feature Policy, HSTS, etc.).

With the new design, there is no such signal, so I can only think of the following possibilities:

- Determine based on `User-Agent` whether the UA supports origin policy, and send current headers if not.
- Always send both the current set of headers and origin policy to all user agents, at least until legacy user agents fade away.

Neither of these is great.

The quick fix here is to put the `Sec-Origin-Policy: 0` header back in. However I wanted to check with some folks before doing so.

In particular this reasoning seems to imply that adding any new HTTP response header that supersedes a previous one, requires pairing it with a feature-detection HTTP request header, and sending that for every request. Maybe adding such headers is rare enough, that HTTP hasn't run into it much in the past, and this is really a special case? @mnot, any thoughts?

**Comment by annevk** [comment]
What does the `0` signify? Support?

Adding a new header to each request isn't great either. Rocky transition phase might not be so bad.

**Comment by mikewest** [comment]
> What does the 0 signify? Support?

It signals nothing more than support.

Chrome used to send `Sec-Origin-Policy: 1`, but some clever Google servers decided it would be a great idea to reflect all the request headers delivered by the client, which meant that they sent `Sec-Origin-Policy: 1`, and we tried to load a policy named `1`, and failed, and therefore blocked access to the server. Hooray for bugs.

**Comment by annevk** [comment]
Given that the response header (now?) has a different name (no `Sec-` prefix), we are no longer constrained by that, but good to know.

**Comment by mikewest** [comment]
Yup. Different names for the request and response headers completely solves that problem.

**Comment by Sora2455** [comment]
Has this issue been resolved? The Chrome engineers are scheduling the release of Origin Policy for the 83/84 release, but without some way to feature detect support, I don't see how anyone can use the feature.

**Comment by domenic** [comment]
@Sora2455 this issue is still labeled "open design question". I'm currently leaning toward @annevk's suggestion of "rocky transition phase might not be so bad", which means you can use it in one of the two ways listed in the OP. However, if experimentation with developers on the flagged implementation reveals that this is too burdensome, we will reassess. After all, it's easier to add something later, than it is to take it away.

**Comment by Sora2455** [comment]
@domenic that it is, and I appreciate there are implementation issues, but unless there is some way to feature detect support for this feature, the only thing developers can use to determine its support is the infamously brittle and unreliable user agent detection.

The second suggestion from the OP I can't see anyone adopting, as it has no benifits (Origin Policy is overridden/duplicated in existing headers).

If this was released tomorrow, and the only way to detect support was client-side, I would use client-side feature detection to set a cookie so that the server would know that it's safe to use Origin Policy - effectively polyfilling the Sec-Origin-Policy header. However, at the moment, I don't believe I can even feature detect this client-side.

**Comment by domenic** [comment]
Thanks, I understand this feature might not work for you until it rolls out to all browsers, and that's valuable feedback.

**Comment by Sora2455** [comment]
Can you set cookies on the well-known/originPolicy response? If so, developers could set the Origin-Policy header on some fraction of responses, then if the cookie is set serve it on every response and not the headers it replaces.

**Comment by domenic** [comment]
Interesting. Currently you cannot, but that is another open design question, in #21. The fact that credentials mode also governs `Set-Cookie` is something I wasn't aware of until now; I'd only been thinking of its impact on requests, not responses.

**Comment by Sora2455** [comment]
Actually, I've just realised there _is_ a way to feature detect Origin Policy client-side (as long as you do not currently have an extensive Feature Policy). If you add a Feature Policy in your Origin Policy restricting, say, midi or USB to www.originPolicy.supported, then you can check for Origin Policy support client-side like so:
`const originPolicyApplied = document.featurePolicy.getAllowlistForFeature("USB").includes("www.originPolicy.supported")`

Mind, this only works if you aren't setting a Feature Policy header with a USB value (as that would override the Origin Policy value).

**Comment by mnot** [comment]
Sorry, didn't catch the @.

Generally we don't have such transitions very often in HTTP; headers just hang around until people feel safe omitting them. Negotiating for that doesn't seem worth the bits on the wire for a single header - especially since request bytes are relatively more precious than response bytes. 

However, here the whole point is to get rid of a LOT of response headers, potentially, so the same reasoning may not apply.

It would be helpful to get a list of the response headers we think we could trigger removal for with such a signal. If it includes CSP, for example, that's a pretty big deal, size-wise.

From a CDN perspective, it would be nice to strip response headers based upon the presence of such a signal.

**Comment by domenic** [comment]
> It would be helpful to get a list of the response headers we think we could trigger removal for with such a signal. If it includes CSP, for example, that's a pretty big deal, size-wise.

Cross-referencing with https://github.com/WICG/origin-policy#configurable-policy-items, it should be something like:

- `Content-Security-Policy`, `Content-Security-Policy-Report-Only`
- `Feature-Policy`, `Feature-Policy-Report-Only`
- `Document-Policy`, `Document-Policy-Report-Only`
- `Large-Allocation` (Mozilla nonstandard)
- `Referrer-Policy`
- `Access-Control-Allow-Origin` and related headers, maybe; the design might end up only supporting a subset.
- `Strict-Transport-Security`
- `Expect-CT`
- `Accept-CH`

However this is further complicated by the fact that different implementations may roll out different policy item implementations via origin policy at different times, so having a browser that supports OP doesn't necessarily mean it supports (for example) Referrer-Policy-via-OP.

It feels like we could go two directions here:

- Build up a complicated feature-detection system, e.g. a request header like `Origin-Policy-Supports: content_security, features, referrer`
- Let site/CDN operators maintain a mapping of `User-Agent` headers to what-the-UA-supports-in-OP.

**Comment by Sora2455** [comment]
Assuming that #48 lands, I was planning to implement Origin Policy like this:

1. By default, send all the headers that Origin Policy would normally cover
2. Client-side, check if self.originPolicyIds exists and if it's empty. If it does, make an AJAX call to an endpoint that sets the Origin Policy header with allowed set to the current policy to trigger a mandatory update.
3. When the AJAX call completes, by my understanding, the new Origin Policy has been applied. Set a cookie listing which policy had been applied. (Could include that in the body of the AJAX call, or use my Feature Policy detection trick as above.)
4. If the cookie exists and is equal to the current policy ID, the server-side code knows it's safe to not send the headers that Origin Policy covers.
5. When the Origin Policy needs to be updated, send Allowed=(null) to clear the previous policy to clients that have the cookie set to a previous ID. Then repeat from step 1.

In my case, I really only care about replacing the Content Security Policy, Referrer Policy and Strict Transport Security headers, which are well-implemented in modern browsers. I doubt that anyone implementing Origin Policy would leave out the first two (maybe HSTS, if only because the spec isn't finalised yet).

However, if Origin Policy's support of those items could be detected client-side, that would allow me to more intelligently replace more of my headers by including that support as a flag in the cookie in the above steps. No User-Agent parsing, and no 'yelling into the void' Origin Policy support as a request header.

(User-Agent parsing is about to become much more difficult anyway, if the folks behind UA Client Hints [have their way](https://wicg.github.io/ua-client-hints/#grease).)

**Comment by domenic** [comment]
> I doubt that anyone implementing Origin Policy would leave out the first two (maybe HSTS, if only because the spec isn't finalised yet).

We don't plan to ship referrer policy-via-OP support in our initial implementation of Chrome.

> (User-Agent parsing is about to become much more difficult anyway, if the folks behind UA Client Hints have their way.)

That doesn't seem accurate. It will in fact get easier, since you'll no longer need to parse, but instead can just get the brand and version directly.

**Comment by mnot** [comment]
Yeah, spamming a bunch of request values to enable removing response values doesn't seem like a good win.

UA sniffing makes me throw up a little bit, but it might be the most practical way forward until things stabilise. The only other thing that comes to mind is for the client to send one or more H2 SETTINGS that explain what headers can be omitted from this conn.

I'm not sure if I like that or not; while it saves bytes, it's also pushing semantics down further than they should go. 

Also, any scheme for stripping these headers has to consider how the responses will be cached; if they're reused downstream, that's a deal killer. The "right" way to do this is sending a request header enumerating the capabilities of the client and then sending that header in `Vary`, but this is getting pretty verbose for the use case.

So I could see this working if you UA sniff (or use a H2 SETTING) and then set `Cache-Control: private` on all affected responses. Not brilliant, but workable.

Alternatively, if the major engines can get alignment and ship this quickly, we could get to the point where none of this is necessary.

**Comment by Sora2455** [comment]
> We don't plan to ship referrer policy-via-OP support in our initial implementation of Chrome.

Whoops, so you don't. Oh well, replacing most of my CSP will still be worth it.

> That doesn't seem accurate. It will in fact get easier, since you'll no longer need to parse, but instead can just get the brand and version directly.

Except that the Sec-CH-UA header, according to section 5.2, the area I linked to, shouldn't send the brand and version but _several_ brands and versions, for the explicit goal of making changing behaviour for one browser based on UA strings impractical.

Say all the major browsers ship UA client hints tomorrow, and Chrome ships Origin Policy the month after. (I doubt this will happen, but bear with me).

If I get '"Chrome"; v="76", "Firefox", v="78"' as my Sec-CH-UA value, is the Firefox or Chrome? Do I send Origin Policy or not? I won't be able to tell.

**Comment by yonixw** [comment]
I hope this is the place, because it looks like you are already considering adding cookies by making credentials mode "include" due to #21.

But I want to share another use case why cookies should be included, which will match this selective policy issue.
Since this issue opened 2 years ago, new privacy data regulations like GDPR require consent from user. Most solutions just store a new cookie named "consent". With this cookie, the server can decide whether to return a policy that allow 3rd party tools or not.
